config.go

```diff
@@ -161,85 +160,123 @@ func (config TGFConfig) String() string {
-		maxDuration := int64(3600)
-		roleRegex := regexp.MustCompile(".*:assumed-role/(.*)/.*")
-		if identity, err := sts.New(session).GetCallerIdentity(&sts.GetCallerIdentityInput{}); err == nil {
-			if matches := roleRegex.FindStringSubmatch(*identity.Arn); len(matches) > 0 {
-				if role, err := iam.New(session).GetRole(&iam.GetRoleInput{RoleName: &matches[1]}); err == nil {
-					maxDuration = *role.Role.MaxSessionDuration
-				}
-			}
-		}
-		var profile string
-		if profile = config.tgf.AwsProfile; profile == "" {
-			if profile = os.Getenv("AWS_PROFILE"); profile == "" {
-				profile = "default"
-			}
-		}
-		if askedForMfa {
-			log.Warningf("Your AWS configuration is set to expire your session in %v. This timeout could not be automatically extended due to the session's MFA",
-				duration)
-		} else {
-			session, err = config.getAwsSession(maxDuration)
-			log.Warningf("Your AWS configuration is set to expire your session in %v (automatically extended to %v)",
-				duration,
-				time.Duration(maxDuration)*time.Second)
+		log.Warningf(
+			"Credentials for current AWS session are set to expire in less than one hour (%s). Will extend to %s.",
+			expiresIn,
+			newDuration,
+		)
+
+		log.Warningf(
+			color.WhiteString("You should consider defining %s in your AWS config profile %s"),
+			color.HiBlueString("duration_seconds = %d", newDuration/time.Second),
+			color.HiBlueString(getPrettyAwsProfileName(*tgfConfig)),
+		)
+
+		shortConfig := config
+		config, err = tgfConfig.getAwsConfig(newDuration)
+		if err != nil {
+			log.Warning("Failed to extend current AWS session, will use the current short duration.", err)
+			config = shortConfig
 		}
+	}
+
+	log.Debug("Caching newly created AWS config for future calls")
+	cachedAwsConfig = config
 
-		log.Warningf(color.WhiteString("You should consider defining %s in your AWS config profile %s"),
-			color.HiBlueString("duration_seconds = %d", maxDuration), color.HiBlueString(profile))
+	return config, nil
+}
+
+func guessAwsMaxAssumeRoleDuration(awsConfig aws.Config) time.Duration {
+	fallback := 1 * time.Hour
+	log.Debugf("Trying to figure out the max duration of an AWS assume role operation (fallback=%s)", fallback)
+
+	roleRegex := regexp.MustCompile(".*:assumed-role/(.*)/.*")
+
+	identity, err := sts.NewFromConfig(awsConfig).GetCallerIdentity(context.TODO(), &sts.GetCallerIdentityInput{})
+	if err != nil {
+		log.Debug("Failed, using fallback:", err)
+		return fallback
 	}
-	if err == nil {
-		cachedSession = session
+
+	matches := roleRegex.FindStringSubmatch(*identity.Arn)
+	if len(matches) == 0 {
+		log.Debug("Failed, using fallback: Current role is not an assumed role")
+		return fallback
 	}
-	return session, err
+
+	role, err := iam.NewFromConfig(awsConfig).GetRole(
+		context.TODO(),
+		&iam.GetRoleInput{
+			RoleName: &matches[1],
+		},
+	)
+	if err != nil {
+		log.Debug("Failed, using fallback:", err)
+		return fallback
+	}
+
```

Comment:
Should we log the found assumed role name or ARN here? I feel that information would be very useful to debug.

Reply:
This code does log a few lines below but there's no mention of ARN. I'll add that so that it's super obvious where we got our information.

Reply:
Done